solr/jetty/src/java/org/apache/solr/jetty/ShrinkManager.java

```diff
@@ -0,0 +1,145 @@
+
+    /**
+     * This is the number of times that CaS will retry. Non-configurable default is to never retry.
+     * This variable only exists to be set in tests to achieve more precise/reliable behavior, and
+     * even at that it should only make a difference for the case where `idleTimeoutDecay > 1`, and
+     * only when there are a handful of threads -- &gt; 1 (enough to generate contention), but not
+     * so many as that another thread would simply pick up the missed CaS "reservation".
+     */
+    static int RETRY_LIMIT = 0;
+
+    public ShrinkManager(int maxSize)
+    {
+        // We make size a power of two for easy mask/circular operation.
+        // NOTE: it's perfectly fine for `head`/`tail` to overflow/wrap around.
+        int size = Integer.highestOneBit(maxSize + 1) << 1;
+        this.mask = size - 1;
+        this.timestamps = new long[size];
+    }
+
+    /**
+     * Called by a thread that is newly idle, either because it just completed a job, or because
+     * it has just been created. Returns <code>true</code>, indicating that an entry has been added
+     * to this {@link ShrinkManager} for the calling thread.
+     */
+    public boolean onIdle()
+    {
+        timestamps[head.getAndIncrement() & mask] = NanoTime.now();
+        return true;
+    }
+
+    /**
+     * Called by a thread just before it performs work. Returns <code>false</code>, indicating that
+     * an entry (corresponding to the calling thread) has been removed from this {@link ShrinkManager}.
+     */
+    public boolean onBusy()
+    {
+        head.getAndDecrement();
+        return false;
+    }
+
+    /**
+     * Check to see if this pool has idle capacity and is eligible to shrink. If this method returns
+     * <code>false</code>, it has no side-effects. If this method returns <code>true</code>, the
+     * {@link ShrinkManager} has been updated to record the removal of the thread, and it is the
+     * responsibility of the caller to ensure that exactly one corresponding thread (the calling
+     * thread) exits, <i>without</i> calling {@link #prune()}.
+     *
+     * @param itNanos time (in nanos) that pool capacity must be idle in order to be eligible to shrink
+     * @param siNanos minimum time (in nanos) between the removal of threads from the pool
+     * @return <code>true</code> if the pool should shrink, otherwise <code>false</code>.
+     */
+    public boolean pollIdleShrink(long itNanos, long siNanos)
+    {
+        long now = NanoTime.now();
+        final long idleBaseline = timestamps[tail.getAndIncrement() & mask];
+        if (NanoTime.elapsed(idleBaseline, now) > itNanos)
+        {
+            long last = lastShrink.get();
+            int retries = 0;
+            while (NanoTime.elapsed(last, now) > siNanos)
+            {
+                // shrinkInterval is satisfied
+                if (lastShrink.compareAndSet(last, Math.max(last, idleBaseline) + siNanos))
```

Comment:
What will happen with following usecase.
1. process is running with min-threads.
2. sudden activity happens and it create N more threads. Now total threads are (min-threads + N)
3.  Lets assume following things happen
3.1 all timestamps has same  idleBaseline (Or 3.1.1 all threads come at same time here)
3.2 all threads will get same idleBaseline, and one will succeed here (most important only one thread will succeed as RETRY_LIMIT = 0?)
3.3 except one thread, all thread will continue in  QTP.run() method and wait for idletimeout for next job?
3.4 then (3.1 - 3.3) again happens after ideltimeout.

Can this cause same behavior as today, where each thread shrinks after idletimeout?

Reply:
The situation you're describing is addressed by [the comment](https://github.com/cowpaths/fullstory-solr/blob/fd3a7b71bc54abcb33e268433971aeb556317a44/solr/jetty/src/java/org/apache/solr/jetty/ShrinkManager.java#L109-L112) immediately below:
```java
                    // NOTE: attempted CaS may fail, _very_ infrequently. If it does, that's fine.
                    // This is a "best effort" approach to shrinking, and even if our CaS fails,
                    // the missed "shrink reservation" will very likely simply be picked up by
                    // another thread.
```
Worst-case scenario here is that an opportunity to shrink is missed. Even in tests where we're specifically trying to trigger this behavior, such as [testPrecise()](https://github.com/cowpaths/fullstory-solr/blob/fd3a7b71bc54abcb33e268433971aeb556317a44/solr/jetty/src/test/org/apache/solr/jetty/QueuedThreadPoolShrinkTest.java#L262-L276), the behavior manifests infrequently. But we are able to trigger it sometimes, and this is the reason for the existence of the adjustable `RETRY_LIMIT` static field: in artificial situations where we care about shrink behavior with extreme precision (really only writing tests I suspect), we need to retry to avoid the race condition you're describing.

But "in real life":
1. situations triggering this issue will be rare, and
2. even if the issue is hit, the more threads, the more likely it is that the "shrink reservation" will be picked up by another thread, and 
3. even in the event we hit this issue in a way that fully manifests, the consequence is minor to the point of being benign